PurchasesCoreSwift/Purchasing/ProductsManager.swift

```diff
@@ -54,10 +54,15 @@ extension ProductsManager: SKProductsRequestDelegate {
     func productsRequest(_ request: SKProductsRequest, didReceive response: SKProductsResponse) {
         queue.async { [self] in
             Logger.rcSuccess(Strings.network.skproductsrequest_received_response)
-            guard let requestProducts = self.productsByRequests[request] else { fatalError("couldn't find request") }
```

Comment:
Since we used to `fatalError` previously and that causes the app to crash, what's the impact of this change on the state of the app if this happens now that the app will continue to live? Is the SDK left in a weird/invalid state? Is there any more recovery that needs to happen?

Reply:
this is a peculiar one, and I'm torn on what's best:

you can make multiple `SKProductsRequest`s for different products. 
these `fatalError`s would be triggered if when we receive the response from the OS, we don't know which request it was for, or which products it was for. 

I can imagine this happening if we have a bug or if the OS has a bug where it calls the delegate more than once, for example. 

In those cases, I'm not 100% sure what the best behavior we can provide is. 

If we get a response but don't have a callback for it, then either: 
- we've already responded to the callback, in which case, things are actually fine and the app can continue
- we've never responded to the callback, and have lost track of it. In this case, the call will hang waiting for the callback, unless there's a timeout built into the call that will cancel it. We currently don't have one, but a developer could. 

in the `fatalError` world, that would instead translate into a crash that cannot be caught. 

In practice, this is only currently being used for calculating intro trial eligibility.

Reply:
in obj-c, we have similar code that handles this with an `NSAssert`

Reply:
Ok, this seems like it's low-risk/low-impact. I'd interested to know if these fatals have been triggered ever, but of course that is something we'd have to be told about.

Reply:
agreed, it's a bummer we don't have a better way of knowing about these. 
Typically we get GH Issues or support tickets when the SDK itself crashes, but of course there's probably a percentage of devs that don't actually report the crash or ask about it and just stop using the SDK or roll back to earlier versions when they experience issues.